lib/gc/Dialect/Linalgx/Utils.cpp

```diff
@@ -0,0 +1,451 @@
+    attr.mPacking = {PackingMap{{1}, {0}}};
+    attr.nPacking = {PackingMap{{2}, {1}}};
+    attr.kPacking = {PackingMap{{2}, {1, 3}}};
+  } break;
+  default:
+    break;
+  }
+  return attr;
+}
+
+//****************************************************************************//
+//                      Vnni Matmul Utils                                     //
+//****************************************************************************//
+
+Value createMatmulCalc(OpBuilder &b, Location loc, ValueRange args) {
+  assert(args.size() == 3 && "Matmul region expects 3 args.");
+  // Get data type
+  auto outTy = args[2].getType();
+  bool isTypeFP = llvm::isa<FloatType>(outTy);
+  bool isTypeInt = llvm::isa<IntegerType>(outTy);
+  auto createMulCalc = [&](Value val0, Value val1) -> Value {
+    if (isTypeFP)
+      return b.create<arith::MulFOp>(loc, val0, val1);
+    if (isTypeInt)
+      return b.create<arith::MulIOp>(loc, val0, val1);
+    return nullptr;
+  };
+  auto createAddCalc = [&](Value val0, Value val1) -> Value {
+    if (isTypeFP)
+      return b.create<arith::AddFOp>(loc, val0, val1);
+    if (isTypeInt)
+      return b.create<arith::AddIOp>(loc, val0, val1);
+    return nullptr;
+  };
+  // Create calc
+  assert((isTypeFP || isTypeInt) && "Matmul must have valid type.");
+  Value value1 = convertScalarToDtype(b, loc, args[0], outTy, false);
+  Value value2 = convertScalarToDtype(b, loc, args[1], outTy, false);
+  Value value3 = createMulCalc(value1, value2);
+  Value value4 = createAddCalc(args[2], value3);
+  return value4;
+}
+
+FailureOr<linalg::GenericOp>
+makeGenericVnniMatmulOp(OpBuilder &builder, Location loc, VnniOpType opType,
+                        ValueRange inputs, ValueRange outputs) {
+  // Check input/output size
+  if (inputs.size() != 2 || outputs.size() != 1) {
+    return emitError("input/output size must be 2/1!");
+  }
+  // Get shapes of inputs and output
+  auto shapeA = cast<ShapedType>(inputs.front().getType());
+  auto shapeB = cast<ShapedType>(inputs.back().getType());
+  auto shapeC = cast<ShapedType>(outputs.back().getType());
+  // Attr of packed matmul
+  auto vnniAttr = getVnniPackingAttr(opType);
+  // Verify dims and shape is valid
+  if (!verifyPacking(shapeA, shapeB, shapeC, vnniAttr) ||
+      !isWeightShapeVnni(shapeB, vnniAttr.weightDims)) {
+    return emitError("Failed to verify vnni packing!");
+  }
+  // Get attrs for GenericOp
+  auto indexingMaps = getIndexingMaps(builder, //
+                                      shapeA, shapeB, shapeC, vnniAttr);
+  auto iteratorTypes = getIteratorTypesArray(vnniAttr);
+  // Make the GenericOp
+  return builder.create<linalg::GenericOp>(
+      loc, shapeC, inputs, outputs, indexingMaps, iteratorTypes,
+      [&](OpBuilder &b, Location loc, ValueRange args) {
+        Value result = createMatmulCalc(b, loc, args);
+        b.create<linalg::YieldOp>(loc, result);
+      });
+}
+
+bool isGenericVnniMatmulOp(Operation *op, VnniOpType opType) {
```

Comment:
For BRGEMM3D, where do we check whether the GenericOp has `batch-reduce` semantic? It seems that a matmul with different semantics on extra dims (other than M, N, K) could also pass the check

Reply:
Added AffineMap Expr check for any loop order.